File: EmojibleIOS/EmojibleIOS/Services/AuthenticationManager.swift
//
//  AuthenticationManager.swift
//  EmojibleIOS
//
//  Created by Haldun Balim on 2.11.2020.
//

import Foundation
import Firebase


class AuthenticationManager {
    let firebaseAuth = Auth.auth()
    var currentUser:User?
    let notificationCenter = NotificationCenter.default

    private init(){
    }
    
    func startListeningForAuthStateChanges(){
        _ = Auth.auth().addStateDidChangeListener { [unowned self] (auth, user) in
            currentUser = Auth.auth().currentUser
            notificationCenter.post(name: .authStateChanged, object: nil)
        }
    }
    
    private static var instance: AuthenticationManager!
    public static func getInstance() -> AuthenticationManager{
        if instance == nil{
            instance = AuthenticationManager()
        }
        return .instance
    }
    
    // TODO: Not Implemented Yet
    public func isUserLoggedIn() -> Bool{
        return firebaseAuth.currentUser != nil
    }
    
}


// email functions
extension AuthenticationManager{
    
    func checkIfUserWithEmailExists(email: String, completion: @escaping (Any?) -> Void){
        Auth.auth().fetchSignInMethods(forEmail: email){ result, error in
            if let error = error {
                completion(error)
            }else{
                completion(result)
            }
        }
    }
    
    
    func createUserWithEmailAndPassword(email: String, password: String, completion: @escaping ((String?) -> Void)){
        Auth.auth().createUser(withEmail: email, password: password) { authResult, error in
            if let error = error{
                completion(error.localizedDescription)
            }else{
                completion(nil)
            }
        }
    }
    
    
    func signInWithEmailAndPassword(email:String, password:String, completion: @escaping ((String?) -> Void)){
        Auth.auth().signIn(withEmail: email, password: password) {user, error in
            if let error = error {
                completion(error.localizedDescription)
            }else{
                completion(nil)
            }
        }
    }
    
    
}

    
// utility functions
extension AuthenticationManager{
    
    func signOut() -> String? {
        do {
            // UserDatabaseManager.singleton.currentUser = nil
            try Auth.auth().signOut()
            // LoginManager().logOut()
            return nil
        } catch let signOutError as NSError {
            return "Error signing out: \(signOutError)"
        }
    }
    
    func resetPassword(email:String, completion: @escaping ((String?) -> Void)) {
        Auth.auth().sendPasswordReset(withEmail: email) { error in
            if let error = error {
                completion(error.localizedDescription)
            }else{
                completion(nil)
            }
        }
    }
   /*
     // May not be needed
    func linkFacebookToAccount(completion: @escaping ((String?) -> Void)){
        if !isLoggedInWithFBAuth{
            completion("Hesabınız Facebook ile bağlanmış durumda")
        }else if let user = Auth.auth().currentUser {
            let credential = FacebookAuthProvider.credential(withAccessToken: AccessToken.current!.tokenString)
            user.link(with: credential) { (authResult, error) in
                if let error = error {
                    completion(error.localizedDescription)
                }else{
                    completion(nil)
                }
            }
        }
    }
    
    func linkPasswordToAccount(password:String, completion: @escaping ((String?) -> Void)){
        if let user = Auth.auth().currentUser {
            let credential = EmailAuthProvider.credential(withEmail:UserDatabaseManager.singleton.currentUser!.email, password: password)
            user.link(with: credential) { (authResult, error) in
                if let error = error {
                    completion(error.localizedDescription)
                }else{
                    completion(nil)
                }
            }
        }
    }
    */
}

File: EmojibleAndroid/app/src/main/java/com/dji/emojibleandroid/services/AuthenticationManager.kt
package com.dji.emojibleandroid.services

import com.google.firebase.auth.*
import com.google.firebase.ktx.Firebase
import java.lang.Exception

// import com.facebook.AccessToken

class AuthenticationManager private constructor() {
    val firebaseAuth = FirebaseAuth.getInstance()
    var currentUser: FirebaseUser? = null
    val notificationCenter = NotificationCenter.instance

    fun startListeningForAuthChanges() {
        firebaseAuth.addAuthStateListener {
            currentUser = FirebaseAuth.getInstance().currentUser
            notificationCenter.post(Changes.authStateChanged, null)
        }
    }

    fun userLoggedIn(): Boolean {
        return firebaseAuth.currentUser != null
    }

    // Email Functions
    fun checkIfUserWithEmailExists(email: String, completion: (Any?) -> Unit) {
        firebaseAuth.fetchSignInMethodsForEmail(email).addOnCompleteListener {
            if (it.isSuccessful) {
                completion(it.result)
            } else {
                completion(it.exception)
            }
        }
    }

    fun createUserWithEmailAndPassword(
        email: String,
        password: String,
        completion: ((String?) -> Unit)
    ) {
        firebaseAuth.createUserWithEmailAndPassword(email, password).addOnCompleteListener {
            if (it.isSuccessful) {
                completion(null)
            } else {
                completion(it.exception?.message)
            }
        }
    }

    fun signInWithEmailAndPassword(
        email: String,
        password: String,
        completion: (String?) -> Unit
    ) {
        firebaseAuth.signInWithEmailAndPassword(email, password).addOnCompleteListener {
            if (it.isSuccessful) {
                completion(null)
            } else {
                completion(it.exception?.message)
            }
        }
    }

    fun signOut(): String? {
        return try {
            firebaseAuth.signOut()
            null
        } catch (e: Exception) {
            "Error signing out ${e.message}"
        }
    }

    fun resetPassword(email: String, completion: (String?) -> Unit) {
        firebaseAuth.sendPasswordResetEmail(email).addOnCompleteListener {
            if (it.isSuccessful) {
                completion(null)
            } else {
                completion(it.exception?.localizedMessage)
            }
        }
    }

//    // Facebook login functions
//
//    fun facebookSignIn(completion: ((String?) -> Void)){
//        //also registers user if there is no user registered with that username
//
//        val credential = FacebookAuthProvider.getCredential(AccessToken.current!.tokenString)
//        //firebaseLogin(credential, completion: completion)
//    }

    private object HOLDER {
        val INSTANCE = AuthenticationManager()
    }

    companion object {
        val instance: AuthenticationManager by lazy { HOLDER.INSTANCE }
    }
}